sdk/core/Azure.Core/src/Pipeline/AzureOperationScope.cs

```diff
@@ -0,0 +1,86 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Diagnostics;
+
+namespace Azure.Core.Pipeline
+{
+    public struct DiagnosticScope: IDisposable
+    {
+        private Activity _activity;
+
+        private readonly string _name;
+
+        private readonly DiagnosticListener _source;
+
+        internal DiagnosticScope(string name, DiagnosticListener source)
+        {
+            _name = name;
+            _source = source;
+            _activity = _source.IsEnabled() ? new Activity(_name) : null;
+        }
+
+        public bool IsEnabled => _activity != null;
+
+        public void AddAttribute(string name, string value)
+        {
+            _activity?.AddTag(name, value);
+        }
+
+        public void AddAttribute<T>(string name, T value)
+        {
+            if (_activity != null)
```

Comment:
I'd avoid doing this. Activity tags are always strings.  ToString() can be called by caller and it's better to make it clear from API.

Reply:
This allows us to avoid formatting to string when there is no diagnostic source attached.

Reply:
there is a special Activity.Recorded flag now - it should be false if there is no listener and if you badly need tags - you should add them only if activity is recorded (i.e. sampling in). OpenTelemetry should take case of setting it to true when needed.

Reply:
Would there be a point in creating activity at all if it's not being recorded?

Reply:
yes, for propagation: you don't record it, but service before you and after you did - they can correlate telemetry (but causation is broken)